From f1c927a0bb4af9a49fe85accce07ddeb23a49cef Mon Sep 17 00:00:00 2001
From: Jack Klika <7218605+jackklika@users.noreply.github.com>
Date: Sun, 4 May 2025 14:08:37 -0500
Subject: [PATCH 2/4] Add unread_thread_notifications to SyncResponse (#540)

* Add unread_thread_notifications to SyncResponse

If `unread_thread_notifications` is passed as part of a sync filter, it will be included in the Sync response. This change allows nio to expose it.
---
 src/nio/responses.py | 43 +++++++++++++++++++++++++++++--------------
 1 file changed, 29 insertions(+), 14 deletions(-)

diff --git a/src/nio/responses.py b/src/nio/responses.py
index d51178f..f3c3d32 100644
--- a/src/nio/responses.py
+++ b/src/nio/responses.py
@@ -253,6 +253,9 @@ class RoomInfo:
     account_data: List = field()
     summary: Optional[RoomSummary] = None
     unread_notifications: Optional[UnreadNotifications] = None
+    unread_thread_notifications: Optional[Dict[str, UnreadNotifications]] = (
+        None  # {event_id: UnreadNotifications}
+    )
 
     @staticmethod
     def parse_account_data(event_dict):
@@ -2003,6 +2006,7 @@ class SyncResponse(Response):
         summary_events: Dict[str, Any],
         unread_notification_events: Dict[str, Any],
         account_data_events: List[Any],
+        unread_thread_notifications: Dict[str, Any],
     ) -> RoomInfo:
         state = SyncResponse._get_room_events(state_events)
 
@@ -2022,15 +2026,23 @@ class SyncResponse(Response):
             unread_notification_events.get("highlight_count"),
         )
 
+        unread_thread_notifications_dict: Dict[str, UnreadNotifications] = {
+            event_id: UnreadNotifications(
+                d.get("notification_count"), d.get("highlight_count")
+            )
+            for event_id, d in unread_thread_notifications.items()
+        }
+
         account_data = RoomInfo.parse_account_data(account_data_events)
 
         return RoomInfo(
-            timeline,
-            state,
-            ephemeral_event_list,
-            account_data,
-            summary,
-            unread_notifications,
+            timeline=timeline,
+            state=state,
+            ephemeral=ephemeral_event_list,
+            account_data=account_data,
+            summary=summary,
+            unread_notifications=unread_notifications,
+            unread_thread_notifications=unread_thread_notifications_dict,
         )
 
     @staticmethod
@@ -2052,14 +2064,17 @@ class SyncResponse(Response):
 
         for room_id, room_dict in parsed_dict.get("join", {}).items():
             join_info = SyncResponse._get_join_info(
-                room_dict.get("state", {}).get("events", []),
-                room_dict.get("timeline", {}).get("events", []),
-                room_dict.get("timeline", {}).get("prev_batch"),
-                room_dict.get("timeline", {}).get("limited", False),
-                room_dict.get("ephemeral", {}).get("events", []),
-                room_dict.get("summary", {}),
-                room_dict.get("unread_notifications", {}),
-                room_dict.get("account_data", {}).get("events", []),
+                state_events=room_dict.get("state", {}).get("events", []),
+                timeline_events=room_dict.get("timeline", {}).get("events", []),
+                prev_batch=room_dict.get("timeline", {}).get("prev_batch"),
+                limited=room_dict.get("timeline", {}).get("limited", False),
+                ephemeral_events=room_dict.get("ephemeral", {}).get("events", []),
+                summary_events=room_dict.get("summary", {}),
+                unread_notification_events=room_dict.get("unread_notifications", {}),
+                account_data_events=room_dict.get("account_data", {}).get("events", []),
+                unread_thread_notifications=room_dict.get(
+                    "unread_thread_notifications", {}
+                ),
             )
 
             joined_rooms[room_id] = join_info
-- 
2.52.0

